clients/src/main/java/org/apache/kafka/common/utils/ImplicitLinkedHashMultiSet.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.common.utils;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * A memory-efficient hash multiset which tracks the order of insertion of elements.
+ * See org.apache.kafka.common.utils.ImplicitLinkedHashSet for implementation details.
+ *
+ * This class is a multi-set because it allows multiple elements to be inserted that are
```

Comment:
Wondering if this should be the exceptional behavior and not the rule. Which APIs currently expect multiset semantics? I realize that there is a difficulty enforcing set semantics when parsing requests, but it might be nice to catch violations when building a response.

Reply:
I think we need multisets here in order to avoid breaking compatibility.  One example is CreateTopics -- the KIP specifies what should happen when there are duplicate elements, and it's not a disconnect.  I would expect a similar story for all of the RPCs, unfortunately.